Fix issues with CSW harvesting

ebuckley opened the pull request:
Opening this up for early review: still a bit of refactoring and planning of next steps to do.

1. syntax errors with catch statment
2. logical issues with
3. re-project the spatial extent if it is in a different srid

ebuckley committed:
initial changes fixes a few mapping issues:

1. syntax errors with catch statment
2. logical issues with
3. re-project the spatial extent if it is in a different srid

ebuckley committed:
remove set_trace() and clean up logging

ebuckley commented:
NOTE: we need to add pyproj to the module now that we are depending on it

anotheredward commented on ckanext/dia/harvester/csw.py:
```diff
@@ -206,6 +210,54 @@ def get_package_dict(self, context, data_dict):
+                except ValueError:
+                    log.warn('Failed to parse json for spatial field of package {}'.format(package_dict))
+
+        if not not spatial_srid and spatial_geojson is not None:
```
I haven't seen `not not` in python before, so that's interesting.
I'm assuming this doesn't work if it's
`if spatial_srid` instead?
`if bool(spatial_srid)` is also an option.

ebuckley replied:
This could probably be a bit clearer. This weird conditional is a result of the following.
- sometimes spatial_srid is the empty string `''`
- the empty_string == False.


I think that `if bool(spatial_srid)` is a refactor that makes sense on this one

anotheredward replied:
I can't seem to work out how:
`if spatial_srid` is  functionally different than `if bool(spatial_srid)` or `if not not spatial_srid`
They're both `False` when `spatial_srid` is `''` or `None`
They're both true when `spatial_srid` has any characters?

I think i'd generally prefer `if spatial_srid`
What am I missing?

As an aside, I know we have `asbool` for when you want a string like `'True'` to convert into the boolean `True`, and other things not so much, it's used a lot in the codebase.

anotheredward reviewed (approved):
LGTM

anotheredward committed:
add polygon support

anotheredward committed:
add the pyproj package as a dependency

anotheredward committed:
add license information

anotheredward committed:
add note on manual re-harvesting

anotheredward committed:
add PR feedback

anotheredward committed:
fix incorrect conversion for already WGS84 geometry

When both CornerPoints and Spatial boundaries are provided
the single SRID provided could be for either
the CornerPoints or the Spatial Boundaries
So we just take the Spatial Boundaries as-is, as in the error
case we're currently seeing the CornerPoints are in EPSG:2193
and need conversion, while the spatial boundaries are already in
WGS84 and it's converting values that are already in WGS84 incorrectly
as if they were in EPSG:2193 that is causing the issue.

anotheredward committed:
pick up additional valid date fields

ebuckley commented on README.md:
````diff
@@ -11,6 +11,13 @@ You can use `pip` to install this plugin into your virtual environment:
+docker-compose run --service-ports ckan-fetch ckan-paster --plugin=ckanext-harvest harvester import -p 'reefton-goldfield-sheet-12-part-of-waitahu-survey-district-and-pohaturoa-survey-district-field-1'
+```
+where `reefton-goldfield...` is the dataset id in the url, this may also look like a guid, both will work.
+
````
This is a sweet tip

ebuckley commented on ckanext/dia/harvester/csw.py:
```diff
@@ -206,6 +230,74 @@ def get_package_dict(self, context, data_dict):
+                    new_extras.append(extra)
+            package_dict['extras'] = new_extras
+        else:
+            log.warning('Could not determine SRID or spatial field for package {}'.format(package_dict))
```
Seems like we could refactor this. logic, as there are a lot of branches to content with at the moment. But I'm largely happy with the logic.

ebuckley reviewed (commented):
Have reviewed, LGTM for release on UAT to test. Some refactoring would be nice to have around the bigger piece of logic we have added in the custom CSW harvest code.
